### PR TITLE
Add XPM support to image library

### PR DESCRIPTION
**Pull request by andydotxyz**
It’s not all versions of XPM as some are more complex than others.
This is a refactoring out of FyneDesk with the additional registration code so it works as a Go image decoder.

### CODE REVIEW

File: xpm/xpm.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+func parseXPM(data []byte) image.Image {
+	var colCount, charSize int
+	colors := make(map[string]color.Color)
+	var img *image.NRGBA
+
+	rowStart := 0
+	rowNum := 0
+	for i, b := range data {
+		if b != '\n' {
```
Comment (Bluebugs) [inline xpm/xpm.go]: I think this code would be a lot cleaner if it was using https://pkg.go.dev/bufio#Scanner and parseXPM could work on a io.Reader I would think.

**Comment by Bluebugs** [comment]
This looks better, I kind of feel it would be safer to handle error during parsing and report them upward to Decode. If the xpm file is downloaded, it could be a problem if we don't handle error detection properly.

**Comment by andydotxyz** [comment]
> This looks better, I kind of feel it would be safer to handle error during parsing and report them upward to Decode. If the xpm file is downloaded, it could be a problem if we don't handle error detection properly.

I could not see any error handing with `Scanner.Scan()` can you see something I missed?

**Comment by Bluebugs** [comment]
There is a few place in the code where errors returned are ignored by using _ and when scanner.Scan() find error, it does count them in scanner.ErrorCount, I think.

**Comment by andydotxyz** [comment]
Thanks for the tips, took a while to hunt down those _. Fixed :)

File: xpm/xpm.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+func parseDimensions(data string) (w, h, i, j int) {
+	if len(data) == 0 {
+		return
+	}
+	parts := strings.Split(data, " ")
+	if len(parts) != 4 {
+		return
+	}
+
+	w, _ = strconv.Atoi(parts[0])
```
Comment (Bluebugs) [inline xpm/xpm.go]: This one also can return an error, right?

File: xpm/xpm.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	if len(parts) == 2 && parts[0] == "c" {
+		parts = []string{" ", "c", parts[1]}
+	} else if len(parts) != 3 {
+		return
+	} else if parts[1] != "c" {
+		return
+	}
+
+	color, err := stringToColor(parts[2])
+	return data[:charSize], color, nil
```
Comment (Bluebugs) [inline xpm/xpm.go]: May have forgotten to return err too.

**Review by Bluebugs — CHANGES_REQUESTED** [review]

File: xpm/xpm.go [hunk 4]
```diff
@@ -72,10 +75,10 @@ func parseDimensions(data string) (w, h, i, j int) {
...
 	}
 
-	w, _ = strconv.Atoi(parts[0])
-	h, _ = strconv.Atoi(parts[1])
-	i, _ = strconv.Atoi(parts[2])
-	j, _ = strconv.Atoi(parts[3])
+	w, err = strconv.Atoi(parts[0])
+	h, err = strconv.Atoi(parts[1])
+	i, err = strconv.Atoi(parts[2])
+	j, err = strconv.Atoi(parts[3])
```
Comment (Bluebugs) [inline xpm/xpm.go:90]: I am not certain here, but wouldn't this code only return an error if it happen during the parsing of parts[3]?

File: xpm/xpm.go (hunk 2 above)
Comment (andydotxyz) [reply xpm/xpm.go]: That one never even crossed my mind, added

File: xpm/xpm.go (hunk 4 above)
Comment (andydotxyz) [reply xpm/xpm.go:90]: I was trying to be clever to avoid error handling code twice as long as the real code :(.
Added the returns.

**Review by Bluebugs — APPROVED** [review]

**Pull request merged**